### PR TITLE
Add metrics for AbstractRoutingDataSource.

### PR DESCRIPTION
**Pull request by bono007**
Fixes gh-22823

### CODE REVIEW

**Comment by bono007** [comment]
I will fix the conflicts w/in next 24 hrs - just noticed it.

File: spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/metrics/jdbc/DataSourcePoolMetricsAutoConfiguration.java [hunk 1]
```diff
@@ -69,10 +74,31 @@
...
+					return;
+				}
+				if (properties.isIgnoreRoutingDataSources()) {
+					return;
+				}
+				AbstractRoutingDataSource routingDataSource = (AbstractRoutingDataSource) dataSource;
+				routingDataSource.getResolvedDataSources().forEach((routingKey, routedDataSource) -> {
+					boolean alreadyRegistered = dataSources.containsValue(routedDataSource);
+					if (!properties.isDeduplicateRoutingDataSources() || !alreadyRegistered) {
+						allDataSources.put(getRoutedDataSourceName(beanName, routingKey), routedDataSource);
```
Comment (wilkinsona) [inline spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/metrics/jdbc/DataSourcePoolMetricsAutoConfiguration.java:96]: This is tricky.

If a `DataSource` is a target of multiple routing data sources, the name for its metrics will depend upon the ordering of the `dataSources` map. Similarly, if a `DataSource` is both a top-level bean and a target of an `AbstractRoutingDataSource` its name will also depend upon the ordering of the `dataSources` map.

I'm not sure exactly what the answer to this is at the moment, but I am sure that the names need to be deterministic. While the injected `dataSources` map is a `LinkedHashMap` (so it'll iterate in insertion order), I'm not sure that Framework guarantees that order. My reading of `DefaultListableBeanFactory.findAutowireCandidates` is that it does not.

Let's take a step back until we've completed the design work on #22823.

**Comment by wilkinsona** [comment]
Thanks again for your efforts here, @bono007. I'm afraid that we've decided that we need to change tack on this one. Please see [this comment](https://github.com/spring-projects/spring-boot/issues/22823#issuecomment-817939467) for details.

**Comment by bono007** [comment]
No worries @wilkinsona - sounds like a good plan.

**Pull request closed without merging**